File: backend/src/main/java/com/trio/backend/service/AnnouncementService.java
package com.trio.backend.service;

import com.trio.backend.dto.announcement.AnnouncementResponse;
import com.trio.backend.dto.announcement.CreateAnnouncementRequest;
import com.trio.backend.dto.announcement.UpdateAnnouncementRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

public interface AnnouncementService {

    AnnouncementResponse create(UUID workspaceId, CreateAnnouncementRequest request);

    AnnouncementResponse getById(UUID workspaceId, UUID announcementId);

    Page<AnnouncementResponse> listWorkspaceAnnouncements(UUID workspaceId, Pageable pageable);

    Page<AnnouncementResponse> listDepartmentAnnouncements(UUID workspaceId, UUID departmentId, Pageable pageable);

    Page<AnnouncementResponse> listTeamAnnouncements(UUID workspaceId, UUID teamId, Pageable pageable);

    Page<AnnouncementResponse> listProjectAnnouncements(UUID workspaceId, UUID projectId, Pageable pageable);

    AnnouncementResponse update(UUID workspaceId, UUID announcementId, UpdateAnnouncementRequest request);

    void delete(UUID workspaceId, UUID announcementId);
}

File: frontend/project/src/services/announcement-service.ts
import { apiClient } from '../lib/api';
import type { PageResponse } from '../types/api';

export interface AnnouncementResponse {
  id: string;
  workspaceId: string;
  departmentId?: string;
  teamId?: string;
  projectId?: string;
  title: string;
  content: string;
  isPinned: boolean;
  status: string;
  createdAt: string;
  updatedAt: string;
  createdBy: string;
}

export interface CreateAnnouncementRequest {
  title: string;
  content: string;
  isPinned?: boolean;
  departmentId?: string;
  teamId?: string;
  projectId?: string;
}

export interface UpdateAnnouncementRequest {
  title?: string;
  content?: string;
  isPinned?: boolean;
}

function base(wsId: string) {
  return `/workspaces/${wsId}/announcements`;
}

export const announcementService = {
  list: (wsId: string) =>
    apiClient.get<PageResponse<AnnouncementResponse>>(`${base(wsId)}`),

  getById: (wsId: string, annId: string) =>
    apiClient.get<AnnouncementResponse>(`${base(wsId)}/${annId}`),

  create: (wsId: string, data: CreateAnnouncementRequest) =>
    apiClient.post<AnnouncementResponse>(`${base(wsId)}`, data),

  update: (wsId: string, annId: string, data: UpdateAnnouncementRequest) =>
    apiClient.put<AnnouncementResponse>(`${base(wsId)}/${annId}`, data),

  delete: (wsId: string, annId: string) =>
    apiClient.delete<void>(`${base(wsId)}/${annId}`),

  listByDepartment: (wsId: string, deptId: string) =>
    apiClient.get<PageResponse<AnnouncementResponse>>(`${base(wsId)}/departments/${deptId}`),

  listByTeam: (wsId: string, teamId: string) =>
    apiClient.get<PageResponse<AnnouncementResponse>>(`${base(wsId)}/teams/${teamId}`),

  listByProject: (wsId: string, projId: string) =>
    apiClient.get<PageResponse<AnnouncementResponse>>(`${base(wsId)}/projects/${projId}`),
};
